src/diffusers/pipelines/audioldm2/modeling_audioldm2.py

```diff
@@ -0,0 +1,1546 @@
+    embedding vectors at the start and end of each text embedding sequence respectively.
+
+    Args:
+        text_encoder_dim (`int`):
+            Dimensionality of the text embeddings from the first text encoder (CLAP).
+        text_encoder_2_dim (`int`):
+            Dimensionality of the text embeddings from the second text encoder (T5).
+        langauge_model_dim (`int`):
+            Dimensionality of the text embeddings from the language model (GPT2).
+    """
+
+    @register_to_config
+    def __init__(self, text_encoder_dim, text_encoder_2_dim, langauge_model_dim):
+        super().__init__()
+        # additional projection layers
+        self.clap_projection = nn.Linear(text_encoder_dim, langauge_model_dim)
+        self.t5_projection = nn.Linear(text_encoder_2_dim, langauge_model_dim)
+
+        # learnable SOS / EOS token embeddings
+        self.clap_sos_embed = nn.Parameter(torch.ones(langauge_model_dim))
+        self.clap_eos_embed = nn.Parameter(torch.ones(langauge_model_dim))
+
+        self.t5_sos_embed = nn.Parameter(torch.ones(langauge_model_dim))
+        self.t5_eos_embed = nn.Parameter(torch.ones(langauge_model_dim))
+
+    def forward(
+        self,
+        clap_hidden_states: Optional[torch.FloatTensor] = None,
+        t5_hidden_states: Optional[torch.FloatTensor] = None,
+        clap_attention_mask: Optional[torch.LongTensor] = None,
+        t5_attention_mask: Optional[torch.LongTensor] = None,
+    ):
+        clap_hidden_states = self.clap_projection(clap_hidden_states)
+        clap_hidden_states, clap_attention_mask = add_special_tokens(
+            clap_hidden_states, clap_attention_mask, sos_token=self.clap_sos_embed, eos_token=self.clap_eos_embed
+        )
+
+        t5_hidden_states = self.t5_projection(t5_hidden_states)
+        t5_hidden_states, t5_attention_mask = add_special_tokens(
+            t5_hidden_states, t5_attention_mask, sos_token=self.t5_sos_embed, eos_token=self.t5_eos_embed
+        )
+
+        # concatenate clap and t5 text encoding
+        hidden_states = torch.cat([clap_hidden_states, t5_hidden_states], dim=1)
+
+        # concatenate attention masks
+        if clap_attention_mask is None and t5_attention_mask is not None:
+            clap_attention_mask = t5_attention_mask.new_ones((clap_hidden_states[:2]))
+        elif clap_attention_mask is not None and t5_attention_mask is None:
+            t5_attention_mask = clap_attention_mask.new_ones((t5_hidden_states[:2]))
+
+        if clap_attention_mask is not None and t5_attention_mask is not None:
+            attention_mask = torch.cat([clap_attention_mask, t5_attention_mask], dim=-1)
+        else:
+            attention_mask = None
+
+        return AudioLDM2ProjectionModelOutput(
+            hidden_states=hidden_states,
+            attention_mask=attention_mask,
+        )
+
+
+class AudioLDM2UNet2DConditionModel(ModelMixin, ConfigMixin, UNet2DConditionLoadersMixin):
```

Comment:
I think this is good if we need to make a separate class because this is a one off pipeline. I think it's possible @patrickvonplaten might have a different opinion but I don't think it's worth blocking because we can merge into the existing unet after the fact if we need to. 

So let's just move forward here and I'll file an issue asking patrick to double check he's ok with it being a separate class when he's back from vacation :)

Reply:
https://github.com/huggingface/diffusers/issues/4658 follow up issue